
--- qtlockedfile.cpp  1970-01-01 01:00:00.000000000
+++ qtlockedfile.cpp  2008/05/16 10:51:19.000000000
@@ -1,5 +1,7 @@
 #include "qtlockedfile.h"
 
+namespace SharedTools {
+
 /*!
     \class QtLockedFile
 
@@ -123,3 +125,5 @@
 
     Destroys the \e QtLockedFile object. If any locks were held, they are released.
 */
+
+}

--- qtlockedfile.h  1970-01-01 01:00:00.000000000
+++ qtlockedfile.h  2008/05/16 10:51:19.000000000
@@ -19,6 +19,8 @@
 #  define QT_QTLOCKEDFILE_EXPORT
 #endif
 
+namespace SharedTools {
+
 class QT_QTLOCKEDFILE_EXPORT QtLockedFile : public QFile
 {
 public:
@@ -41,4 +43,6 @@
     LockMode m_lock_mode;
 };
 
+}
+
 #endif

--- qtlockedfile_unix.cpp  1970-01-01 01:00:00.000000000
+++ qtlockedfile_unix.cpp  2008/05/16 10:51:19.000000000
@@ -5,6 +5,8 @@
 
 #include "qtlockedfile.h"
 
+namespace SharedTools {
+
 bool QtLockedFile::lock(LockMode mode, bool block)
 {
     if (!isOpen()) {
@@ -73,3 +75,4 @@
         unlock();
 }
 
+}

--- qtlockedfile_win.cpp  1970-01-01 01:00:00.000000000
+++ qtlockedfile_win.cpp  2008/05/16 10:51:19.000000000
@@ -2,6 +2,8 @@
 #include <qt_windows.h>
 #include <QtCore/QFileInfo>
 
+namespace SharedTools {
+
 #define SEMAPHORE_PREFIX "QtLockedFile semaphore "
 #define MUTEX_PREFIX "QtLockedFile mutex "
 #define SEMAPHORE_MAX 100
@@ -168,3 +170,4 @@
     }
 }
 
+}
