diff --git a/src/plugins/qt4projectmanager/qt4projectmanager.pro b/src/plugins/qt4projectmanager/qt4projectmanager.pro
index 317544c..f9ba1d1 100644
--- a/src/plugins/qt4projectmanager/qt4projectmanager.pro
+++ b/src/plugins/qt4projectmanager/qt4projectmanager.pro
@@ -144,7 +144,6 @@ FORMS += makestep.ui \
     wizards/mobileappwizardmaemooptionspage.ui \
     librarydetailswidget.ui \
     qtversioninfo.ui \
-    debugginghelper.ui \
     debugginghelper.ui
 RESOURCES += qt4projectmanager.qrc \
     wizards/wizards.qrc
